### ISSUE TITLE
Early stopping in NN training

### ISSUE THREAD

**Issue by bjschoenfeld** [byu-dml/d3m-dynamic-neural-architecture#98]
Often neural networks reach a point in their training where they stop improving and may even get worse. We can detect that this is happening by looking at the validation loss/scores. We could have these models stop their training early to avoid wasted effort and get results sooner. We could have a parameter that indicates if a model has not improved in the last k epochs of training, it should stop.

**Comment by bjschoenfeld** [comment]
We have "early stopping" implemented in that the models train for `n_epochs` and then use the weights corresponding to the lowest validation error. We are using our test data as validation data for this purpose, which means our test data is being used to help determine the model's weights. This is a big no-no in machine learning.

Rather, the model should reserve a portion of the training data to use as validation data to determine the stopping point. This does mean that the model has less training data, but better generalization performance. Is the cost worth the benefit?
